data/transactions/logic/eval.go

```diff
@@ -3648,10 +3908,66 @@ func opAppParamsGet(cx *EvalContext) {
 	cx.stack = append(cx.stack, stackValue{Uint: exist})
 }
 
+func opAcctParamsGet(cx *EvalContext) {
+	last := len(cx.stack) - 1 // acct
+
+	if cx.Ledger == nil {
+		cx.err = fmt.Errorf("ledger not available")
+		return
+	}
+
+	addr, _, err := cx.accountReference(cx.stack[last])
+	if err != nil {
+		cx.err = err
+		return
+	}
+
+	paramField := AcctParamsField(cx.program[cx.pc+1])
+	fs, ok := acctParamsFieldSpecByField[paramField]
+	if !ok || fs.version > cx.version {
+		cx.err = fmt.Errorf("invalid acct_params_get field %d", paramField)
+		return
+	}
+
+	bal, err := cx.Ledger.Balance(addr)
+	if err != nil {
+		cx.err = err
+		return
+	}
+	exist := boolToUint(bal.Raw > 0)
+
+	var value stackValue
+
+	switch fs.field {
+	case AcctBalance:
+		value.Uint = bal.Raw
+	case AcctMinBalance:
+		mbal, err := cx.Ledger.MinBalance(addr, cx.Proto)
+		if err != nil {
+			cx.err = err
+			return
+		}
+		value.Uint = mbal.Raw
+	case AcctAuthAddr:
+		auth, err := cx.Ledger.Authorizer(addr)
+		if err != nil {
+			cx.err = err
+			return
+		}
+		if auth == addr {
```

Comment:
that's.. unexpected. If the `AuthAddr` on the account data happen to be the same as the public key address, why would you use the empty address ?

Reply:
The `Authorizer()` method's job is to report the correct authorizer for the account, regardless of whether it is the original key, or new key that has been set by rekey.  But the job of this field is to report the AuthAddr - they key that an account has been rekeyed to if it it has been rekeyed at all.  So, when the Authorizer() reports the original key, there is no AuthAddr, and the field returns the zero address.

Note that this is not ambiguous.  You can not set the AuthAddr of an account to the original public key.  If you try, it simply gets unset.
```
// Rekey updates tx.Sender's AuthAddr to tx.RekeyTo, if provided
func Rekey(balances Balances, tx *transactions.Transaction) error {
	if (tx.RekeyTo != basics.Address{}) {
		acct, err := balances.Get(tx.Sender, false)
		if err != nil {
			return err
		}
		// Special case: rekeying to the account's actual address just sets acct.AuthAddr to 0
		// This saves 32 bytes in your balance record if you want to go back to using your original key
		if tx.RekeyTo == tx.Sender {
			acct.AuthAddr = basics.Address{}
		} else {
			acct.AuthAddr = tx.RekeyTo
		}

		err = balances.Put(tx.Sender, acct)
		if err != nil {
			return err
		}
	}
	return nil
}
```

Reply:
Oh I see what you did here. Admittedly, I think that it's even worse. We shouldn't make the evaluator depend on the semantics of setting the AuthAddr field on the AccountData. What we really want is to have access to the AccountData.AuthAddr here, right ? I think that adding this to LedgerForLogic would work out nicely here. The COW should already have that cached as part of the basics.AccountData.

Reply:
Yes, it would certainly be worse if I had introduced `Authorizer` for this purpose. Why would I combine the public key with the authaddr, just to figure that out and tear them apart again later? Perhaps there's more going on.

Authorizer() was introduced for authorization in v5. When checking if an inner transaction is allowed, the Authorizer() of the intended Sender is compared to the app address. If they are the same, the transaction is authorized. This keeps the mechanics of how authaddr overrides the public key out of logic evaluation.

Next, and separately, in v6 there was a desire to expose the AuthAddr to teal code. I could have introduced yet another method to LogicForLedger, but it seems crazy to have two methods intended to deal with getting info about the rekey address. So, I left the existing Authorizer() method alone (so that authentication code could remain less complicated) and put complexity in the "safe" place - exposing the authaddr to teal access.

So to summarize. Rekey info is needed in two places in evaluation - authorization and field exposure. Since authorization came first and is intuitively more important to get right, I left the code alone and put the complexity in field access.

If I switch to exposing the authaddr directly in LedgerForLogic, instead of offerring Authorizer() as a method, I have to repeat the code to override the private key with the authaddr in the logic evaluator's authorization code. I can't see the win there.

Reply:
I think that the implementation you've made for `Authorizer` is fine, and there is no need to change it. It's just the new `AuthAddr` field that is the issue here. When retrieving this field, we should avoid any logic being applied and just provide it as is. Having the extra logic would mean that the correctness of the teal logic would be tied to the implementation of the rekey apply logic; that's a design risk that could be easily avoided.

Reply:
Ok, I can add AccountData() as a LedgerForLogic methods, and remove Balance() and MinBalance(). Then get all of that info from the AccountData.

One begins to wonder if the LedgerForLogic interface is doing much any more, as we've had to expose so much for c2c and this.